source/common/common/cleanup.h

```diff
@@ -16,17 +18,33 @@ class Cleanup {
 };
 
 // RAII helper class to add an element to an std::list on construction and erase
-// it on destruction.
-template <class T> class ListAddAndRemove {
+// it on destruction, unless the cancel method has been called.
+template <class T> class RaiiListElement {
 public:
-  ListAddAndRemove(std::list<T>& container, T element) : container_(container) {
+  RaiiListElement(std::list<T>& container, T element) : container_(container), cancelled_(false) {
     it_ = container.emplace(container.begin(), element);
   }
-  virtual ~ListAddAndRemove() { container_.erase(it_); }
+  virtual ~RaiiListElement() {
+    if (!cancelled_) {
+      erase();
+    }
+  }
+
+  // Cancel deletion of the element on destruction. This should be called if the iterator has
+  // been invalidated, eg. if the list has been cleared or the element removed some other way.
+  void cancel() { cancelled_ = true; }
+
+  // Delete the element now, instead of at destruction.
+  void erase() {
+    ASSERT(!cancelled_);
+    container_.erase(it_);
```

Comment:
Can you add a unit test for this? Also needs master merge.

Reply:
Done